README.md
# Laravel Flash Message

[![Latest Version](https://img.shields.io/github/release/ubient/laravel-flash-message.svg?style=flat-square)](https://github.com/ubient/laravel-flash-message/releases)
[![Build Status](https://img.shields.io/travis/ubient/laravel-flash-message/master.svg?style=flat-square)](https://travis-ci.org/ubient/laravel-flash-message)
[![Quality Score](https://img.shields.io/scrutinizer/g/ubient/laravel-flash-message.svg?style=flat-square)](https://scrutinizer-ci.com/g/ubient/laravel-flash-message)
[![StyleCI](https://styleci.io/repos/154986115/shield)](https://styleci.io/repos/154986115)
[![Total Downloads](https://img.shields.io/packagist/dt/ubient/laravel-flash-message.svg?style=flat-square)](https://packagist.org/packages/ubient/laravel-flash-message)

This package provides a Laravel Flash Message that can be used to display an one-time status message to your users.

## Usage

When redirecting, you can fluently chain one of the [available methods](#available-methods), similar to how you would [flash other session data](https://laravel.com/docs/7.x/redirects#redirecting-with-flashed-session-data):

```php
return redirect()
    ->route('login')
    ->withErrorMessage('To execute this action, you need to be logged in first.');
```

With the message flashed to the session, all that remains is to display it to your users.
A good place to start doing so, is by including the alert we've shipped with this package into your view(s):
```php
@include('flash-message::alert')
```

If you're testing your application, and would like to assert whether the flash message was set, you can do so
by calling one of the [available assertions](#available-assertions) on the TestResponse object:
```php
// @see https://laravel.com/docs/7.x/http-tests#testing-json-apis
$testResponse = $this->post('/topics/1/replies', ['message' => 'My example reply']);

// Assert that the success flash message was set, and that it contains the expected message:
$testResponse->assertHasSuccessMessage('Your reply has been added.');

// Alternatively, we can also assert that some (any) success flash message was set:
$testResponse->assertHasSuccessMessage();
```

#### Available Methods

- `withInfoMessage('Message')`: Flashes a message that indicates a neutral informative change or action.
- `withSuccessMessage('Message')`: Flashes a message that indicates a successful or positive action.
- `withWarningMessage('Message')`: Flashes a message that indicates a warning that might need attention.
- `withErrorMessage('Message')`: Flashes a message that indicates an erroneous, dangerous or negative action.

#### Available Assertions

- `assertHasInfoMessage('Message')`: Asserts that a flash message was set using `withInfoMessage`.
- `assertHasSuccessMessage('Message')`: Asserts that a flash message was set using `withSuccessMessage`
- `assertHasWarningMessage('Message')`: Asserts that a flash message was set using `withWarningMessage`
- `assertHasErrorMessage('Message')`: Asserts that a flash message was set using `withErrorMessage`

### Customizing the template
By default, our alert uses [TailwindCSS](https://github.com/tailwindcss/tailwindcss)'s utility classes to style the alert,
but odds are you're either using something else or want to modify how the alert looks altogether.
Luckily, doing so is fairly easy:

1. Publish our template into your application by running the following artisan command:
```bash
php artisan vendor:publish --provider="Ubient\FlashMessage\FlashMessageServiceProvider"
```
2. Modify the template located at `resources/views/vendor/flash-message/alert.blade.php`.
3. That's all! Laravel will automatically pick up on the changes you've made.


## Installation

You can install the package via composer:

```bash
composer require ubient/laravel-flash-message
```

The package will automatically register itself.

## Testing

You can run the test using:

``` bash
vendor/bin/phpunit
```

## Upgrading

Please see [UPGRADING](UPGRADING.md) for details.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email claudio@ubient.net instead of using the issue tracker.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `ubient/laravel-flash-message` will be documented in this file

## 5.1.0 - 2020-09-07
- Add support for Laravel 8

## 5.0.0 - 2020-04-12
- Add support for Laravel 7.0
- Fixed a crash when not in an environment where testing-related packages installed.
- Upgrade TailwindCSS classes to 1.0

## 4.0.0 - 2019-09-03
- Drop support for Laravel 5.7 and older
- Add support Laravel 6.0

## 3.0.0 - 2019-02-27
- Drop support for PHP 7.1
- Add support for Laravel 5.8

## 2.0.2 - 2019-02-01
- Fixed an incorrect namespacing issue for multiple files in the `tests/Unit/` folder
- Removed (opinionated) linter overrides from tests

## 2.0.1 - 2018-11-04
- Remove a trailing semicolon in `alert.blade.php`

## 2.0.0 - 2018-11-04
- [Persist flash messages across redirects - #8](https://github.com/ubient/laravel-flash-message/issues/8)

## 1.1.0 - 2018-11-03
- Add ability to assert flash messages

## 1.0.0 - 2018-10-29

- Initial release


tests/TestCase.php
<?php

namespace Ubient\FlashMessage\Tests;

use LogicException;
use Orchestra\Testbench\TestCase as Orchestra;
use Ubient\FlashMessage\FlashMessageServiceProvider;

class TestCase extends Orchestra
{
    /**
     * @param \Illuminate\Foundation\Application $app
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            FlashMessageServiceProvider::class,
        ];
    }

    /**
     * Detect and return the correct TestResponse class.
     *
     * @return string
     */
    protected function getTestResponse(): string
    {
        // Laravel >= 7.0
        if (class_exists("\Illuminate\Testing\TestResponse")) {
            return "\Illuminate\Testing\TestResponse";
        }

        // Laravel <= 6.0
        if (class_exists("\Illuminate\Foundation\Testing\TestResponse")) {
            return "\Illuminate\Foundation\Testing\TestResponse";
        }

        throw new LogicException('Could not find TestResponse class.');
    }

    /**
     * Whether or not the TestResponse class exists.
     *
     * @return bool
     */
    protected function hasTestResponseClass(): bool
    {
        try {
            $this->getTestResponse();
        } catch (LogicException $exception) {
            return false;
        }

        return true;
    }
}


tests/Unit/FlashMessageTest.php
<?php

namespace Ubient\FlashMessage\Tests\Unit;

use Ubient\FlashMessage\FlashMessage;
use Ubient\FlashMessage\Tests\TestCase;

class FlashMessageTest extends TestCase
{
    /** @test */
    public function has_a_levels_method(): void
    {
        $expected = ['info', 'success', 'warning', 'error'];
        $actual = FlashMessage::levels();

        $this->assertEquals($expected, $actual);
    }

    /** @test */
    public function has_an_assert_method(): void
    {
        $flashData = ['level' => 'success', 'message' => 'Example Message'];
        session()->flash('flash_message', $flashData);

        FlashMessage::assert($flashData['level'], $flashData['message']);
    }

    /** @test */
    public function has_a_set_method(): void
    {
        $flashData = ['level' => 'success', 'message' => 'Example Message'];

        FlashMessage::set($flashData['level'], $flashData['message']);

        $this->assertEquals($flashData, session()->get('flash_message'));
        $this->assertEquals($flashData, session()->get('flash_message'));
    }
}


tests/Feature/ViewFlashMessageTest.php
<?php

namespace Ubient\FlashMessage\Tests\Feature;

use PHPUnit\Framework\ExpectationFailedException;
use Ubient\FlashMessage\Tests\TestCase;

class ViewFlashMessageTest extends TestCase
{
    const CSS_CLASS_INFO = 'bg-blue-400';

    const CSS_CLASS_SUCCESS = 'bg-green-400';

    const CSS_CLASS_WARNING = 'bg-orange-500';

    const CSS_CLASS_ERROR = 'bg-red-400';

    public function setUp(): void
    {
        parent::setUp();

        app('router')->group(['middleware' => 'web'], function ($router) {
            $router->get('display', function () {
                return view('flash-message::alert');
            });
        });
    }

    /** @test */
    public function it_should_display_our_info_message(): void
    {
        $message = 'A wide screen just makes a bad film twice as bad.';

        redirect('display')->withInfoMessage($message);
        $response = $this->get('display');

        $response->assertSeeText($message);
        $response->assertSee(static::CSS_CLASS_INFO);
        $response->assertDontSee(static::CSS_CLASS_ERROR);
    }

    /** @test */
    public function it_should_display_our_success_message(): void
    {
        $message = 'Where there is no vision, there is no hope.';

        redirect('display')->withSuccessMessage($message);
        $response = $this->get('display');

        $response->assertSeeText($message);
        $response->assertSee(static::CSS_CLASS_SUCCESS);
        $response->assertDontSee(static::CSS_CLASS_INFO);
    }

    /** @test */
    public function it_should_display_our_warning_message(): void
    {
        $message = 'Nine-tenths of wisdom is being wise in time.';

        redirect('display')->withWarningMessage($message);
        $response = $this->get('display');

        $response->assertSeeText($message);
        $response->assertSee(static::CSS_CLASS_WARNING);
        $response->assertDontSee(static::CSS_CLASS_SUCCESS);
    }

    /** @test */
    public function it_should_display_our_error_message(): void
    {
        $message = 'Life is the art of drawing without an eraser.';

        redirect('display')->withErrorMessage($message);
        $response = $this->get('display');

        $response->assertSeeText($message);
        $response->assertSee(static::CSS_CLASS_ERROR);
        $response->assertDontSee(static::CSS_CLASS_WARNING);
    }

    /** @test */
    public function it_should_be_backwards_compatible_with_laravels_built_in_status_message(): void
    {
        $message = 'These are used by default in ResetsPasswords.php and SendsPasswordResetEmails.php';

        redirect('display')->with('status', $message);
        $response = $this->get('display');

        $response->assertSeeText($message);
        $response->assertSee(static::CSS_CLASS_SUCCESS);
    }

    /** @test */
    public function it_should_prefer_our_flashed_message_over_laravels_built_in_status_message(): void
    {
        $message = "All I ask is the chance to prove that money can't make me happy.";
        $builtInMessage = 'These are used by default in ResetsPasswords.php and SendsPasswordResetEmails.php';

        redirect('display')
            ->withWarningMessage($message)
            ->with('status', $builtInMessage);

        $response = $this->get('display');

        $response->assertSeeText(e($message), false);
        $response->assertSee(static::CSS_CLASS_WARNING);
        $response->assertDontSee(e($builtInMessage));
        $response->assertDontSee(static::CSS_CLASS_SUCCESS);
    }

    /** @test */
    public function it_should_display_nothing_when_no_message_was_flashed(): void
    {
        $response = $this->get('display');

        $response->assertViewIs('flash-message::alert');
        $this->assertEmpty($response->content());
    }

    /** @test */
    public function it_should_persist_across_multiple_redirects(): void
    {
        $message = 'Those who do not remember the past are condemned to repeat it.';
        app('router')->group(['middleware' => 'web'], function ($router) use ($message) {
            $router->get('redirect-to-display', function () {
                return redirect('display');
            });
            $router->get('initial', function () use ($message) {
                return redirect('redirect-to-display')->withInfoMessage($message);
            });
        });

        $this->get('initial');
        $response = $this->get('redirect-to-display');

        $response->assertHasInfoMessage($message);
    }

    /** @test */
    public function it_should_only_be_displayed_once(): void
    {
        $message = 'If you want to kill any idea in the world, get a committee working on it.';

        redirect('display')->withInfoMessage($message);
        $this->get('display');
        $response = $this->get('display');

        $this->expectException(ExpectationFailedException::class);
        $response->assertHasInfoMessage($message);
    }
}


src/FlashMessage.php
<?php

namespace Ubient\FlashMessage;

use LogicException;
use PHPUnit\Framework\Assert;

class FlashMessage
{
    /**
     * A list of all supported "flash message" levels.
     *
     * @return array
     */
    public static function levels(): array
    {
        return [
            'info',
            'success',
            'warning',
            'error',
        ];
    }

    /**
     * Flashes the given "flash message" to the session.
     *
     * @param  string  $level
     * @param  string  $message
     * @return void
     */
    public static function set(string $level, string $message): void
    {
        session()->put('flash_message', [
            'message' => $message,
            'level' => $level,
        ]);
    }

    /**
     * Asserts that the "flash message" is set as expected.
     *
     * @param  string  $level
     * @param  string|null  $value
     * @param  string  $message
     * @return void
     */
    public static function assert(string $level, string $value = null, string $message = ''): void
    {
        if (! class_exists("\PHPUnit\Framework\Assert")) {
            throw new LogicException('Could not assert: PHPUnit is not installed or is of incompatible version.');
        }

        Assert::assertThat(
            session()->get('flash_message'),
            new HasFlashMessageConstraint($level, $value),
            $message
        );
    }
}


src/HasFlashMessageConstraint.php
<?php

namespace Ubient\FlashMessage;

use Illuminate\Support\Arr;
use PHPUnit\Framework\Constraint\Constraint;
use SebastianBergmann\Diff\Differ;

class HasFlashMessageConstraint extends Constraint
{
    /**
     * @var string
     */
    private $expectedLevel;

    /**
     * @var string|null
     */
    private $expectedMessage;

    public function __construct(string $expectedLevel, string $expectedMessage = null)
    {
        $this->expectedLevel = $expectedLevel;
        $this->expectedMessage = $expectedMessage;
    }

    /**
     * Returns a string representation of the constraint.
     */
    public function toString(): string
    {
        return 'has a flash message';
    }

    /**
     * Evaluates the constraint for parameter $other. Returns true if the
     * constraint is met, false otherwise.
     *
     * @param  mixed  $other  value or object to evaluate
     * @return bool
     */
    protected function matches($other): bool
    {
        return $this->hasValidFlashData($other)
            && $this->hasExpectedLevel($other)
            && $this->hasExpectedMessage($other);
    }

    /**
     * Returns the description of the failure.
     *
     * The beginning of failure messages is "Failed asserting that" in most
     * cases. This method should return the second part of that sentence.
     *
     * @param  mixed  $other  evaluated value or object
     * @return string
     */
    protected function failureDescription($other): string
    {
        $message = sprintf('an [%s] flash message is set', $this->expectedLevel);

        if ($this->expectedMessage) {
            return $message.' with the expected value';
        }

        return $message;
    }

    protected function additionalFailureDescription($other): string
    {
        if (! $this->hasValidFlashData($other) || ! $this->hasExpectedLevel($other)) {
            return '';
        }

        $differ = new Differ("--- Expected\n+++ Actual\n");

        return $differ->diff($this->expectedMessage, $other['message']);
    }

    /**
     * Checks whether valid "flash message" data was passed.
     *
     * @param $other
     * @return bool
     */
    private function hasValidFlashData($other): bool
    {
        return is_array($other) && Arr::has($other, ['level', 'message']);
    }

    /**
     * Checks whether the "flash message" level is what we expected.
     *
     * @param $other
     * @return bool
     */
    private function hasExpectedLevel($other): bool
    {
        return $other['level'] === $this->expectedLevel;
    }

    /**
     * Checks whether we expect a message, and if so, whether
     * the "flash message" message is what we expected.
     *
     * @param $other
     * @return bool
     */
    private function hasExpectedMessage($other): bool
    {
        if ($this->expectedMessage === null) {
            return true;
        }

        return $other['message'] === $this->expectedMessage;
    }
}


src/FlashMessageServiceProvider.php
<?php

namespace Ubient\FlashMessage;

use Illuminate\Http\RedirectResponse;
use Illuminate\Support\Facades\App;
use Illuminate\Support\ServiceProvider;
use LogicException;

class FlashMessageServiceProvider extends ServiceProvider
{
    public function boot(): void
    {
        $this->registerMacros();

        $this->loadViewsFrom(__DIR__.'/views', 'flash-message');

        $this->publishes([
            __DIR__.'/views' => base_path('resources/views/vendor/flash-message'),
        ]);
    }

    /**
     * Register all "flash message" macros per alert level.
     *
     * @return void
     */
    private function registerMacros(): void
    {
        collect(FlashMessage::levels())->each(function ($level) {
            $this->registerFlashMessage($level);

            if (App::runningUnitTests() && $this->hasTestResponseClass()) {
                $this->registerHasFlashMessageAssertion($level);
            }
        });
    }

    /**
     * Register the "flash message" macro for the given
     * "alert level" on the redirect response.
     *
     * @param  string  $level
     * @return void
     */
    protected function registerFlashMessage(string $level): void
    {
        $name = 'with'.ucfirst(strtolower($level)).'Message';

        RedirectResponse::macro($name, function ($message) use ($level) {
            FlashMessage::set($level, $message);

            return $this;
        });
    }

    /**
     * Register the "flash message" assertion macro for
     * the given "alert level" on the Test Response.
     *
     * @param  string  $level
     * @return void
     */
    protected function registerHasFlashMessageAssertion(string $level): void
    {
        $name = 'assertHas'.ucfirst(strtolower($level)).'Message';

        /** @var \Illuminate\Foundation\Testing\TestResponse|\Illuminate\Testing\TestResponse $testResponse */
        $testResponse = $this->getTestResponse();
        $testResponse::macro($name, function (string $value = null, string $message = '') use ($level) {
            FlashMessage::assert($level, $value, $message);

            return $this;
        });
    }

    /**
     * Detect and return the correct TestResponse class.
     *
     * @return string
     */
    protected function getTestResponse(): string
    {
        // Laravel >= 7.0
        if (class_exists("\Illuminate\Testing\TestResponse")) {
            return "\Illuminate\Testing\TestResponse";
        }

        // Laravel <= 6.0
        if (class_exists("\Illuminate\Foundation\Testing\TestResponse")) {
            return "\Illuminate\Foundation\Testing\TestResponse";
        }

        throw new LogicException('Could not find TestResponse class.');
    }

    /**
     * Whether or not the TestResponse class exists.
     *
     * @return bool
     */
    protected function hasTestResponseClass(): bool
    {
        try {
            $this->getTestResponse();
        } catch (LogicException $exception) {
            return false;
        }

        return true;
    }
}
